### PR TITLE
presession kx

### PR DESCRIPTION
**Pull request by martelletto**
the ~~last~~ first seven commits in this branch implement pre-session kx, post-kx account creation as discussed.

### CODE REVIEW

**Comment by martelletto** [comment]
Please note that this PR introduces an important change: kx is now the first thing that happens when zkclient connects to zkserver. Before this PR, it was possible for zkclient to issue a few commands without authenticating itself (and without authenticating the server).

These commands included rpc.InitialCmdIdentify and rpc.InitialCmdCreateAccount: rpc.InitialCmdIdentify caused zkserver to reply with its public key, and rpc.InitialCmdCreateAccount allowed an account to be created.

With an identity directory in zkserver (which is effectively what this PR implements), we want to ensure that account creation happens after zkclient and zkserver have authenticated each other, thus the need for kx to happen first.

**Comment by martelletto** [comment]
#57

- what it does?

it implements zkserver support for a directory where identities can be
looked up.

- how does it work?

when the branch was first pushed, 3 commands were implemented: /push,
/find, and /pull. /push pushed a user's identity to the directory, /find
<nick> looked <nick> up on the directory, and /pull removed a user's
identity from the directory.

after feedback from marco, the decision to store a user's identity in
the directory was moved from zkclient to zkserver. there's now a
"directory = yes/no" toggle in zkserver.conf (off by default) and it's
up to zkclient to terminate the connection if it does not want its
identity listed in the directory.

- what's missing?

a dialog to warn a connecting zkclient that zkserver has directory = yes
in its zkserver.conf, and to prompt the user to make a decision whether
to proceed with the connection or not.

File: zkclient/zkclient.go [hunk 1]
```diff
@@ -1211,6 +1229,26 @@ func (z *ZKC) handleRPC() {
...
+			}
+			if r.Error != "" {
+				z.PrintfT(0, "find failed: %v", r.Error)
+			} else {
+				z.PrintIdentity(r.Identity)
+			}
+			err = z.tagStack.Push(message.Tag)
+			if err != nil {
+				exitError = fmt.Errorf("IdentityFindReply "+
+					"invalid tag: %v", message.Tag)
```
Comment (dajohi) [inline zkclient/zkclient.go]: do you want to include the actual error returned from Push()?

File: zkserver/account/account.go [hunk 2]
```diff
@@ -151,6 +155,117 @@ func (a *Account) Create(pid zkidentity.PublicIdentity, force bool) error {
...
+				user.Unlock()
+				return id, nil
+			}
+		}
+		user.Unlock()
+		a.Lock()
+	}
+	a.Unlock()
+
+	return nil, fmt.Errorf("user not found")
```
Comment (dajohi) [inline zkserver/account/account.go]: user %s not found?

**Pull request merged**